design_notes/XDN11 - Error handling.md

```diff
@@ -0,0 +1,194 @@
+
+```
+interface IXlangErrorValue : IUnknown
+{
+	void GetValue(xlang_string * errorValue)
+}
+```
+
+A projection can also use this parameter to store other information that it may want to retrieve later or expose to tooling.  An example use of this is to store a language exception or a custom error info.  Lets say an exception thrown in a component propagates through several different components.  One of these components may be written in the same language as the component which threw the exception.  In that case, a projection may want to re-throw the same exception initially thrown.  It can achieve this by storing the information necessary to re-construct the exception in an COM object and passing it to this parameter.  The object passed to this parameter will be stored along with the other error information and will propagate with the error.  It can later be retrieved when the error propagates back to the same language and used to re-construct the exception.  Similarly, a projection which allows to call WinRT APIs may want tooling to be able to access the custom error info (IRestrictedErrorInfo) it has for the error.  It can achieve this by passing the IRestrictedErrorInfo to this parameter and then let tooling query for it from xlang's error info upon failure.
+
+## Projecting the error and associating error info
+
+Upon receiving an error for a call, consuming projections should map the xlang error to the equivalent native error for the language they project.  This will allow for consumers to be written using natural and familiar code for the language they are written in.  But in some cases, there might not be an equivalent native error to map to.  In these cases, it is left to the projection to decide which native error to map to because a projection implementer will be able to make a better decision on what a consumer of the language they project would expect.
+
+Consuming projections should also retrieve the error info associated with the xlang error and associate it in some way with the native error.  It should be associated in a way that allows the logging of the native error to also log the information in the error info.  For instance, in certain languages, exceptions have a ToString function which consumers of that language call upon failure to log the error for later analysis.  This ToString function should also return the information in xlang's error info to allow the developer to get a better picture of the failure.  In addition, it should also be associated in a way that allows a producing projection for the same language to determine if there is an error info associated with the native error and to be able to retrieve it if there is.  This will allow the producing projection to propagate the error with the previously collected diagnostic information if the error ends up being propagated by the component instead of being handled.
```

Comment:
Feels like we're defining a bunch of rules for a projection to be Good.

Maybe pull out the top-line items into a "Guidance for Projection Authors" section, that has things like:

* **DO** invoke `xlang_originate_error` when detecting an error during a call to the implementation
* **DO** invoke `xlang_propagate_error` when passing along an error
* **CONSIDER** providing a projection-specific extended error information that implements `IXLangErrorExtendedInfo`
* **DO NOT** allow language-specific exceptions to leak through the projection

Might help to show a typical call sequence of {C# > projection > projection > Python} and call out the responsibilities of each element in the chain.

Reply:
Yes, that sounds like a good idea.  I added into a guidance section at the top.  Will also have an upcoming change with the diagram.